File: server/transfer.c
/*
 * transfer.c - transfer core
 *
 * Written by
 *  Christian Vogelgsang <chris@vogelgsang.org>
 *
 * This file is part of dtv2ser.
 * See README for copyright notice.
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
 *  02111-1307  USA.
 *
 */

#include <stdint.h>
#include <util/crc16.h>

#include "board.h"

#include "transfer.h"
#include "display.h"
#include "util.h"
#include "timer.h"
#include "param.h"
#include "sertrans.h"

#define min(a,b) ((a<b)?(a):(b))

// the transfer state
transfer_state_t transfer_state;

// the dtv transfer state
dtv_transfer_state_t dtv_transfer_state;

// current host transfer funcs
host_transfer_funcs_t *current_host_transfer_funcs = 0;
// current dtv transfer block func
dtv_transfer_block_func_t current_dtv_transfer_block_func = 0;

// ----- main transfer loop -----

static uint8_t transfer_begin(uint8_t mode,uint32_t length)
{
  // reset transfer state
  transfer_state.length  = 0;
  transfer_state.ms_time = 0;
  transfer_state.result  = TRANSFER_OK;

  // begin host transfer
  uint8_t result = current_host_transfer_funcs->begin_transfer(length);
  if(result!=TRANSFER_OK) {
    transfer_state.result = result;
    return result;
  }

  // reset dtv state
  dtv_transfer_state.transfer_length = 0;
  dtv_transfer_state.mode = mode;

  led_transmit_on();
  return TRANSFER_OK;
}

static uint8_t transfer_end(uint8_t result,uint32_t total_length,uint16_t ms_time)
{
  led_transmit_off();

  // end host transfer
  result = current_host_transfer_funcs->end_transfer(result);

  // set transfer result
  transfer_state.length  = total_length;
  transfer_state.ms_time = ms_time;
  transfer_state.result  = result;

  return result;
}

uint8_t transfer_mem(uint8_t mode,uint32_t base,uint32_t length,uint16_t block_size)
{
  uint8_t result = transfer_begin(mode,length);
  if(result!=TRANSFER_OK)
    return result;

  // start timer
  uint32_t start = timer_now();

  // block copy loop
  uint32_t total_length = 0;
  uint8_t toggle = 1;
  while(length) {
    uint16_t offset = (uint16_t)(base & 0x3fff);
    uint16_t max_len = 0x4000 - offset;
    uint16_t block_len = min(length,block_size);
    uint16_t len = min(block_len,max_len);

    dtv_transfer_state.offset = offset;
    dtv_transfer_state.bank   = (uint8_t)(base >> 14);
    dtv_transfer_state.length = len;
    dtv_transfer_state.crc16  = 0xffff;

    // call dtv func to transfer a single block
    // (calls host_funcs transfer_byte)
    result = current_dtv_transfer_block_func();
    if(result!=TRANSFER_OK)
      break;

    // host check block
    result = current_host_transfer_funcs->check_block(dtv_transfer_state.crc16);
    if(result!=TRANSFER_OK)
      break;

    // update
    uint16_t transfer_length = dtv_transfer_state.transfer_length;
    base   += transfer_length;
    length -= transfer_length;
    total_length += transfer_length;

    // toggle led
    toggle = toggle^1;
    if(toggle) {
      led_transmit_on();
    } else {
      led_transmit_off();
    }

    // break if not ok
    if(result!=TRANSFER_OK)
      break;
  }

  return transfer_end(result,
                      total_length,
                      (timer_now() - start)/10);
}

uint8_t transfer_mem_block(uint8_t mode,uint8_t bank,uint16_t offset,uint16_t length)
{
  uint8_t result = transfer_begin(mode,length);
  if(result!=TRANSFER_OK)
    return result;

  dtv_transfer_state.bank   = bank;
  dtv_transfer_state.offset = offset;
  dtv_transfer_state.length = length;

  // trigger a single block transfer
  result = current_dtv_transfer_block_func();

  return transfer_end(result,
                      dtv_transfer_state.transfer_length,
                      0);
}

// ----- diagnose transfer functions ----------------------------------------
#ifdef USE_DIAGNOSE

uint8_t diagnose_dtv_send_block(void)
{
  uint8_t result;
  uint16_t length = dtv_transfer_state.length;

  uint16_t pos = 0;
  uint16_t crc16 = dtv_transfer_state.crc16;
  while(length) {
    // transfer byte from host
    uint8_t data;
    result = current_host_transfer_funcs->transfer_byte(&data);
    if(result!=TRANSFER_OK)
      break;

    // update checksum
    crc16 = _crc16_update(crc16,data);

    length--;
    pos++;
  }
  dtv_transfer_state.transfer_length = pos;
  dtv_transfer_state.crc16 = crc16;

  return TRANSFER_OK;
}

uint8_t diagnose_dtv_recv_block(void)
{
  uint8_t result;
  uint16_t length = dtv_transfer_state.length;
  uint8_t pattern = PARAM_BYTE(PARAM_BYTE_DIAGNOSE_PATTERN);

  uint16_t pos = 0;
  uint16_t crc16 = dtv_transfer_state.crc16;
  while(length) {
    // transfer pattern byte to host
    result = current_host_transfer_funcs->transfer_byte(&pattern);
    if(result!=TRANSFER_OK)
      break;

    // update checksum
    crc16 = _crc16_update(crc16,pattern);

    pos++;
    length--;
  }
  dtv_transfer_state.transfer_length = pos;
  dtv_transfer_state.crc16 = crc16;

  return TRANSFER_OK;
}

// ----- diagnose host transfer funcs -----

uint8_t diagnose_host_mode = 0; // 0=read, 1=write, 2=verify

static uint8_t diagnose_begin_transfer(uint32_t length)
{ return TRANSFER_OK; }

static uint8_t diagnose_end_transfer(uint8_t lastStatus)
{ return lastStatus; }

static uint8_t diagnose_check_block(uint16_t crc16)
{ return TRANSFER_OK; }

static uint8_t diagnose_transfer_byte(uint8_t *data)
{
  uint8_t pattern = PARAM_BYTE(PARAM_BYTE_DIAGNOSE_PATTERN);

  if(diagnose_host_mode==DIAGNOSE_HOST_MODE_READ) {
    // read from host
    *data = pattern;
  }
  else {
    // write to host
    if(*data != pattern)
      return TRANSFER_ERROR_VERIFY_MISMATCH;
  }
  // do nothing in DIAGNOSE_HOST_MODE_WRITE
  return TRANSFER_OK;
}

host_transfer_funcs_t diagnose_host_transfer_funcs =
{
  .begin_transfer = diagnose_begin_transfer,
  .end_transfer   = diagnose_end_transfer,
  .check_block    = diagnose_check_block,
  .transfer_byte  = diagnose_transfer_byte
};

#endif // USE_DIAGNOSE

File: client/dtv2ser/transfer.py
#
# transfer.py - send and receive data
#
# Written by
#  Christian Vogelgsang <chris@vogelgsang.org>
#
# This file is part of dtv2ser.
# See README for copyright notice.
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
#  02111-1307  USA.
#

import time

from dtv2ser.status  import *
from dtv2ser.cmdline import CmdLine
from dtv2ser.sercon  import SerCon
from dtv2ser.param   import Param

class Transfer:
  """Transfer command handling tools."""

  # how long to wait for the server after a transfer operation
  server_ready_timeout = 500

  # how many retries to get the transfer result after a transfer operation
  get_result_retries = 20

  def __init__(self,con):
    """Create a new transfer object."""
    # get cmdline
    self.con = con
    self.cmdline = CmdLine(con)
    # init transfer rate
    self.server_tx_rate = 0
    self.server_rx_rate = 0
    self.client_tx_rate = 0
    self.client_rx_rate = 0

  # ---------- block size ---------------------------------------------------

  def set_block_size(self,block_size):
    """Set block_size parameter
    Returns the result code
    """
    param = Param(self.con)
    print "\t[set new blocksize: %04x]" % block_size
    return param.set_word(PARAM_WORD_DTV_TRANSFER_BLOCK_SIZE,block_size)

  def get_block_size(self):
    """Get the block_size parameter
    Returns (result,block_size)
    """
    param = Param(self.con)
    return param.get_word(PARAM_WORD_DTV_TRANSFER_BLOCK_SIZE)

  def ensure_block_size(self,block_size):
    """Make sure the given size is set in the dtv2ser device
    Returns result
    """
    result,has_size = self.get_block_size()
    if result != STATUS_OK:
      return result
    if has_size != block_size:
      return self.set_block_size(block_size)
    else:
      return STATUS_OK

  # ---------- transfer mode ------------------------------------------------

  def set_transfer_mode(self,mode):
    """Issue a 'm' command to set the transfer mode
    Returns the result code
    """
  #print "\t[transfer mode=%2d]" % mode
    return self.cmdline.do_command("m%02x" % mode)

  # ---------- transfer result ----------------------------------------------

  def get_transfer_result(self):
    """Issue a 't' command to get the result
    Returns result code and transfer time
    """
    result = self.cmdline.do_command("t")
    if result != STATUS_OK:
      return (result,0)

    # get transfer result
    (result,transfer_result) = self.cmdline.get_byte()
    if result != STATUS_OK:
      return (result,0)
    elif transfer_result != STATUS_OK:
      return (transfer_result | TRANSFER_ERROR_MASK,0)

    # get transfer time
    (result,transfer_time) = self.cmdline.get_word()
    if result != STATUS_OK:
      return (result,transfer_time)
    else:
      return (STATUS_OK,transfer_time)

  def wait_for_transfer_result(self):
    """Wait until the transfer result is available or a time out occurs
    Returns result code and transfer time
    """

    # wait a bit for the result
    for x in xrange(self.get_result_retries):
      result,transfer_time = self.get_transfer_result()
      # everything is ok... return
      if result == STATUS_OK:
        return (STATUS_OK,transfer_time)
      # a transfer error occured -> report it!
      if result & TRANSFER_ERROR_MASK == TRANSFER_ERROR_MASK:
        return (result,time)
      # any other error: retry
      time.sleep(0.1)

    # client timed out
    return (CLIENT_ERROR_SERVER_TIMEOUT,0)

  # ---------- transfer tools -----------------------------------------------

  def calc_crc16(self,data):
    """Calc a block checksum with crc16.
    Return crc16.
    """
    crc = 0xffff
    for a in data:
      b = ord(a)
      crc ^= b;
      for i in xrange(8):
        if crc & 1:
          crc = (crc >> 1) ^ 0xA001
        else:
          crc = (crc >> 1)
    return crc & 0xffff

  def update_server_tx_rate(self,length,server_time):
    """Get transfer time from server and calculate the tx rate.
    Return result.
    """
    if server_time > 0:
      self.server_tx_rate = length * 100.0 / (server_time * 1024.0)
    else:
      self.server_tx_rate = 0

  def update_server_rx_rate(self,length,server_time):
    """Get transfer time from server and calc transfer rate
    Return result.
    """
    if server_time > 0:
      self.server_rx_rate = length * 100.0 / (server_time * 1024.0)
    else:
      self.server_rx_rate = 0

  def update_client_tx_rate(self,length,duration):
    """Update the server transfer rate from the
    """
    if duration != 0:
      self.client_tx_rate = length / (duration * 1024.0)
    else:
      self.client_tx_rate = 0

  def update_client_rx_rate(self,length,duration):
    """After a receive transfer update statistics
    """
    if duration != 0:
      self.client_rx_rate = length / (duration * 1024.0)
    else:
      self.client_rx_rate = 0

  # ---------- block transfer commands --------------------------------------

  def do_send_boot(self,start,data,callback=lambda x:True):
    """Transmit via boot protocol"""
    # send command
    cmd = "b%04x%04x" % (start,len(data))
    result = self.cmdline.do_command(cmd)
    if result != STATUS_OK:
      return result

    result,duration = self.con.send_boot(start,data,callback)
    if result != STATUS_OK:
      return result

    # check transfer result
    result,server_time = self.wait_for_transfer_result()
    if result != STATUS_OK:
      return result

    # update tx rate
    length = len(data)
    self.update_client_tx_rate(length,duration)
    self.update_server_tx_rate(length,server_time)

    return STATUS_OK

  def do_send_command(self,cmd,start,data,callback=lambda x:True,block_size=0x400,mode=TRANSFER_MODE_NORMAL):
    """Transmit a send data command and upload data.
    Return result.
    """
    # send transfer mode
    result = self.set_transfer_mode(mode)
    if result != STATUS_OK:
      return result

    # ensure block size
    result = self.ensure_block_size(block_size)
    if result != STATUS_OK:
      return result

    # send command
    result = self.cmdline.do_command(cmd)
    if result != STATUS_OK:
      return result

    # upload data
    (result,duration) = self.con.send_block(start,data,block_size,callback=callback)
    if result != STATUS_OK:
      return result

    # check transfer result
    result,server_time = self.wait_for_transfer_result()
    if result != STATUS_OK:
      return result

    # update tx rate
    length = len(data)
    self.update_client_tx_rate(length,duration)
    self.update_server_tx_rate(length,server_time)

    return STATUS_OK

  def do_receive_command(self,cmd,start,length,callback=lambda x:True,block_size=0x400,mode=TRANSFER_MODE_NORMAL):
    """Transmit a receive data command and download data.
    Return (result,data)
    """
    # send transfer mode
    result = self.set_transfer_mode(mode)
    if result != STATUS_OK:
      return (result,'')

    # ensure block size
    result = self.ensure_block_size(block_size)
    if result != STATUS_OK:
      return (result,'')

    # send command
    result = self.cmdline.do_command(cmd)
    if result != STATUS_OK:
      return (result,'')

    # download data
    (result,duration,data) = self.con.receive_block(start,length,block_size,callback=callback)
    if result != STATUS_OK:
      return (result,data)

    # check transfer result
    result,server_time = self.wait_for_transfer_result()
    if result != STATUS_OK:
      return (result,data)

    # update rx rate
    self.update_client_rx_rate(length,duration)
    self.update_server_rx_rate(length,server_time)

    return (STATUS_OK,data)

  # ---------- diagnose commands --------------------------------------------

  def set_diagnose_pattern(self,pattern):
    """Set the diagnose pattern in dtv2ser.
    Return result.
    """
    param = Param(self.con)
    #print "\t[set diagnose pattern=%02x]" % pattern
    return param.set_byte(PARAM_BYTE_DIAGNOSE_PATTERN,pattern)

  def get_diagnose_pattern(self):
    """Read the diagnose pattern from dtv2ser.
    Return (result,pattern).
    """
    param = Param(self.con)
    return param.get_byte(PARAM_BYTE_DIAGNOSE_PATTERN)

  def do_diagnose_command_only_dtv(self,cmd,length,is_tx,callback=lambda x:True,block_size=0x400):
    """Issue a diagnose send command but do not transfer data via serial.
    Return result.
    """
    # send transfer mode
    result = self.set_transfer_mode(TRANSFER_MODE_DTV_ONLY)
    if result != STATUS_OK:
      return result

    # ensure block size
    result = self.ensure_block_size(block_size)
    if result != STATUS_OK:
      return result

    # send command
    result = self.cmdline.do_command(cmd);
    if result != STATUS_OK:
      return result

    # wait for result
    start_time = time.time()
    result,server_time = self.wait_for_transfer_result()
    end_time = time.time()
    if result != STATUS_OK:
      return result

    # update tx or rx rate
    if is_tx:
      self.update_client_tx_rate(length,end_time - start_time)
      self.update_server_tx_rate(length,server_time)
    else:
      self.update_client_rx_rate(length,end_time - start_time)
      self.update_server_rx_rate(length,server_time)

    return STATUS_OK

  def do_diagnose_send_command_only_client(self,cmd,start,length,callback=lambda x:True,block_size=0x400):
    """Issue a diagnose send command but do not transfer to the DTV.
    Return result.
    """
    # get pattern
    status,pattern = self.get_diagnose_pattern()
    if status != STATUS_OK:
      return status

    # create test data
    data = chr(pattern) * length

    # perform send command
    return self.do_send_command(cmd,start,data,callback=callback,mode=TRANSFER_MODE_SERIAL_ONLY,block_size=block_size)

  def do_diagnose_receive_command_only_client(self,cmd,start,length,callback=lambda x:True,block_size=0x400):
    """Issue a diagnose receive command but do not transfer to the DTV.
    Return result.
    """
    # get pattern
    status,pattern = self.get_diagnose_pattern()
    if status != STATUS_OK:
      return status

    # perform receive command
    (result,data) = self.do_receive_command(cmd,start,length,callback=callback,mode=TRANSFER_MODE_SERIAL_ONLY,block_size=block_size)
    if result != STATUS_OK:
      return result

    # check data
    expected_data = chr(pattern) * length
    if expected_data != data:
      print "data mismatch:",data
      return CLIENT_ERROR_CORRUPT_SERIAL_DATA
    else:
      return STATUS_OK

  # ---------- get statistics -----------------------------------------------

  def begin_rx_rates(self):
    """Begin a rx operation"""
    self.rx_begin_time = time.time()

  def begin_tx_rates(self):
    """Begin a tx operation"""
    self.tx_begin_time = time.time()

  def get_rx_rates(self,length):
    """End rx operation and return client and server rx rates."""
    t = time.time() - self.rx_begin_time
    return (self.client_rx_rate,self.server_rx_rate,t,length)

  def get_tx_rates(self,length):
    """End tx operation and return client and server tx rates."""
    t = time.time() - self.tx_begin_time
    return (self.client_tx_rate,self.server_tx_rate,t,length)

  # ---------- joy stream ---------------------------------------------------

  def do_joy_stream(self,stream,callback=lambda x:True):
    """Send a joy stream given in string stream
       Returns status,duration"""
    result = self.cmdline.do_command("j")
    if result != STATUS_OK:
      return result

    return self.con.send_joy_stream(stream,callback)

  def do_joy_stream_seq(self,stream_seq,callback=lambda x:True):
    """Send a sequence of joy streams
       Returns status."""
    total_duration = 0
    for js in stream_seq:
      cmd = js[0]

      # send a joy stream
      if cmd == JoyStream_Commands:
        result = self.cmdline.do_command("j")
        if result != STATUS_OK:
          return result,0
        result,duration = self.con.send_joy_stream(js[1],callback)
        if result != STATUS_OK:
          return result,0
        total_duration += duration
      # sleep some seconds
      elif cmd == JoyStream_Sleep:
        duration = js[1]
        time.sleep(duration)
      else:
        print "UNKNOWN Joy Stream Command"

    return STATUS_OK,total_duration
